src/app/modules/submission/multirespondent-submission/__tests__/multirespondent-submission.controller.spec.ts

```diff
@@ -43,17 +54,639 @@ describe('multirespondent-submission.controller', () => {
+      const yesNoFieldId2 = new ObjectId().toHexString()
+
+      const submissionResponses = {
+        [emailFieldId1]: {
+          fieldType: BasicField.Email,
+          answer: {
+            value: expectedEmails[0],
+          },
+        },
+        [emailFieldId2]: {
+          fieldType: BasicField.Email,
+          answer: {
+            value: 'not_expected_1@example.com',
+          },
+        },
+        [yesNoFieldId1]: {
+          fieldType: BasicField.YesNo,
+          answer: 'Yes',
+        },
+        [yesNoFieldId2]: {
+          fieldType: BasicField.YesNo,
+          answer: 'No',
+        },
+      }
+
+      const threeStepApprovalWorkflow: FormWorkflowStepDto[] = [
+        {
+          _id: stepOneId,
+          workflow_type: WorkflowType.Dynamic,
+          field: emailFieldId1,
+          edit: [],
+        },
+        {
+          _id: stepTwoId,
+          workflow_type: WorkflowType.Static,
+          emails: ['not_expected_2@example.com'],
+          edit: [yesNoFieldId1],
+          approval_field: yesNoFieldId1,
+        },
+        {
+          _id: stepThreeId,
+          workflow_type: WorkflowType.Dynamic,
+          field: emailFieldId2,
+          edit: [yesNoFieldId2],
+          approval_field: yesNoFieldId2,
+        },
+      ]
+
+      MockFormService.checkFormSubmissionLimitAndDeactivateForm = jest
+        .fn()
+        .mockReturnValue(okAsync(mockMrfForm))
+
+      MultiRespondentSubmission.findById = jest.fn().mockReturnValue({
+        save: () => true,
+      })
+
+      const mockReq = expressHandler.mockRequest({
+        params: {
+          formId: mockFormId,
+          submissionId: mockSubmissionId,
+        },
+        body: {} as any,
+      })
+      const mockSubmitMrfReq = merge(mockReq, {
+        formsg: {
+          formDef: {
+            _id: mockFormId,
+            authType: FormAuthType.NIL,
+            getUniqueMyInfoAttrs: jest.fn().mockReturnValue([]),
+            workflow: threeStepApprovalWorkflow,
+            emails: [expectedEmails[1]],
+            stepsToNotify: [stepOneId],
+          },
+          encryptedPayload: {
+            encryptedContent: 'encryptedContent',
+            version: 1,
+            submissionPublicKey: 'submissionPublicKey',
+            encryptedSubmissionSecretKey: 'encryptedSubmissionSecretKey',
+            responses: submissionResponses,
+            workflowStep: threeStepApprovalWorkflow.length - 1, // last step
+          },
+        } as any,
+      })
+      const mockRes = expressHandler.mockResponse({})
+
+      // Act
+      await updateMultirespondentSubmissionForTest(mockSubmitMrfReq, mockRes)
+      // Assert
+      // approval email is sent instead of completion email
+      expect(sendMrfApprovalEmailSpy).toHaveBeenCalledTimes(1)
+      expect(sendMrfWorkflowCompletionEmailSpy).not.toHaveBeenCalled()
+      expect(sendMRFWorkflowStepEmailSpy).not.toHaveBeenCalled()
```

Comment:
Did a bit of rabbit-holing on investigating this. I noticed that we're spying through to `MailService` which is something that is usually called by `XService`[1] rather than `XController`.

What triggered this investigating was the difficulty of mocking the spies. As you may have noticed, `updateMultirespondentSubmissionForTest` doesn't actually calls what `sendMRFWorkflowStepEmail` was spying on, but through another interim function on the controller level `sendNextStepEmail`.

Arranging the control flows: 

1. `XController.updateMultirespondentSubmission -> XService.sendMrfOutcomeEmails -> MailService.XXX`
2. `XController.updateMultirespondentSubmission -> XController.sendNextStepEmail -> MailService.sendMRFWorkflowStepEmail`

**Suggestion**
1. Move `XController.sendNextStepEmail` to `XService.sendNextStepEmail`.
2. Update spies over to monitor `XService` instead of `MailService`
3. Spying of MailService can be done on `XService.spec`

**Outcome**
Generally less flakey tests since we're not testing across two different boundaries, i.e., From `Controller -> IntermediaryFoo -> Actual Spy` to `Controller -> Actual Spy`

[1] with some exception such as verification OTP, even if so it can also be argued that the `LoginService` should be calling instead of `LoginController`.

Reply:
Also identified several additional code smell (which I'm also guilty of) [here](https://linear.app/ogp/issue/FRM-1858/encrypt-submissioncontrollerts-should-not-call-save). Something that we should address later on.

Want to just align on the code smell and the impact of such flows.

Reply:
will fix this in a sec!

Reply:
Hi Ken, I've spent some time thinking about and understanding the concerns raised. Let me address them: 

- Regarding spying through to `MailService` being a code smell: 
I think this is a debate between the tradeoffs of [sociable vs solitary unit testing](https://medium.com/clarityai-engineering/sociable-or-solitary-unit-tests-choose-your-tradeoffs-7ced14d4baef).
<img width="727" alt="image" src="https://github.com/user-attachments/assets/d33d811a-5f0f-444e-956a-38eb6f33b65f">
In this case, the unit tests above are meant to be sociable unit tests, where the controller is the interface and we are testing the business functionality as a whole, treating the impl details as a black box.  

To elaborate, here is my justification for advocating for sociable unit tests over writing per layer solitary unit tests for Approvals: 
1. Emulates acceptance tests and less prone to change: Implementation details may change due to refactoring (eg, what service layer function is invoked and arguments etc), however, business functionality remains the same and is the main behavior we are able to assert to prevent regressions.  
2. Save developer time: The above sociable UT are sufficient to cover all acceptance tests. It saves a lot of time (thinking about other features as well) and prevents regressions. 
3. May be more intuitive from another code reader's perspective to understand what each TC's purpose since it relates to acceptance tests: For eg, `sends approved outcome email when mrf is approved for last step of multiple step MRF` rather than `passes isApproval true and isRejected true to sendOutcome` might be easier to understand. 

- Difficulty in mocking spies: 
Spies may be mocked by doing `sendMrfWorkflowCompletionEmailSpy.mockImpl(<impl here>)`. It might be similar even if we changed to solitary UT. 

- Regarding test flakiness: 
I believe you may be referring to the TC failing in the event the code is changed. 
This is a strong justification for writing sociable UT which assert business outcome over per layer unit tests. 
The business functionality is unlikely to change (unless the PRD changes, which TC changes are then inevitable either way). Writing the integration tests above asserts no regressions to functionality while avoiding need to change due to implementation detail changes. 

Therefore, it might not be needed to split into 2 layers and test them solitarily for the reasons above. Please let me know your thoughts on this.

Reply:
If we think about it further, the spying through to `MailService` might not actually be the problem of the test case, but rather the `multirespondent.controller` itself which also partakes in invoking business logic such as `submission.save()` and checking if the step is rejected etc. 

If a change is desired, a more appropriate fix might be to refactor the controller and shift all business logic to the service layer, then apply the written sociable UT to the service layer instead. 

The controller should only contain code which asserts the payload is valid (`ensurePipeline`) before invoking the service function.